### PR TITLE
#62181: postfix.set_master - ignore lines starting with whitespace in…

### LINKED ISSUE #62181: [BUG] postfix.set_master breaks on line continuation
**Description**
When reading a postfix master.cf that has continued lines, set_master gets a line that doesn't match its expectations (ie 8 fields). Lines beginning with '#' are correctly ignored, but lines starting with whitespace should be ignored here as well. As per postfix docs at https://www.postfix.org/master.5.html:
```
SYNTAX
       The general format of the [master.cf](https://www.postfix.org/master.5.html) file is as follows:

       o      Empty  lines and whitespace-only lines are ignored, as are lines
              whose first non-whitespace character is a `#'.

       o      A logical line starts with  non-whitespace  text.  A  line  that
              starts with whitespace continues a logical line.

       o      Each  logical  line defines a single Postfix service.  Each ser-
              vice is identified by its name  and  type  as  described  below.
              When multiple lines specify the same service name and type, only
              the last one is remembered.  Otherwise, the order  of  [master.cf](https://www.postfix.org/master.5.html)
              service definitions does not matter.

       Each  logical  line  consists  of eight fields separated by whitespace.
       These are described below in the order as they appear in the  [master.cf](https://www.postfix.org/master.5.html)
       file.

       Where  applicable  a  field  of  "-" requests that the built-in default
       value be used. For boolean fields specify "y" or "n"  to  override  the
       default value.
```

**Setup**
Install postfix on ubuntu 20.04, try to use postfix.set_master. The file contains continuation lines which dont contain 8 "columns" eg:
```
...
smtp      unix  -       -       -       -       -       smtp
relay     unix  -       -       -       -       -       smtp
        -o syslog_name=postfix/$service_name
#       -o smtp_helo_timeout=5 -o smtp_connect_timeout=5
showq     unix  n       -       -       -       -       showq
...
```
  
**Steps to Reproduce the behavior**
```
salt-call postfix.set_master smtps inet n y n n 100 smtpd
# index out of range error
```

**Expected behavior**
postfix master config is updated

**Versions Report**
Salt Version:
          Salt: 3004.1
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: 0.31.0
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.8.10 (default, Mar 15 2022, 12:22:08)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2
 
System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.4.0-120-generic
        system: Linux
       version: Ubuntu 20.04 focal

### PR DESCRIPTION
**Pull request by HG00**
… master.cf

### What does this PR do?
Removes lines starting with whitespace from consideration in python.set_master

### What issues does this PR fix or reference?
Fixes: #62181

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by MKLeb** [comment]
Hi @HG00 and thanks for your contribution.  A couple of questions...

- What is the purpose of ignoring lines that start with whitespace?
- Would it be better to maybe use regular expressions to check this so we don't have to list all of the possible whitespace characters manually?

Also, this will need tests and a changelog in order for this to get merged.  If you have any questions, don't hesitate to ask.

**Comment by HG00** [comment]
Hi @MKLeb 
_Would it be better to maybe use regular expressions to check this so we don't have to list all of the possible whitespace characters manually?_
I added this as a shortcut for string.whitespace to save an import, and thought it might be a bit quicker and more readable. Let me know if you prefer re and I can swap 

Will work on the test/changelog - sorry should have read the contrib guide!

**Comment by MKLeb** [comment]
@HG00 any updates here?  I do think `re` may be the way to go to detect whitespace as well.

**Comment by HG00** [comment]
Hey @MKLeb sorry this one dropped to the bottom :)
NP with using regex, but there's a choice to make. At the moment the module doesnt care about "flags" lines (which are by default indented). What do you think of adding a "flags" key to the dict that is made from parsing the master.cf? The alternative is we ignore them, but this could break existing installs

**Comment by MKLeb** [comment]
> Hey @MKLeb sorry this one dropped to the bottom :) NP with using regex, but there's a choice to make. At the moment the module doesnt care about "flags" lines (which are by default indented). What do you think of adding a "flags" key to the dict that is made from parsing the master.cf? The alternative is we ignore them, but this could break existing installs

From what I can tell by reading the docs, it looks like anything can be in (what they are calling) the continuation of a logical line, it doesn't have to be flags, at least from the start.  What I'm thinking is that we just continue parsing each logical line until we hit the end of all "flag lines".  Then we continue on as salt already does.  I will leave a suggested change for this, although it may take a bit.

File: salt/modules/postfix.py [hunk 1]
```diff
@@ -55,24 +55,36 @@ def _parse_master(path=MASTER_CF):
...
+                    "unpriv": comps[3],
+                    "chroot": comps[4],
+                    "wakeup": comps[5],
+                    "maxproc": comps[6],
+                    "command": " ".join(comps[7:]),
+                }
+                dict_key = "{} {}".format(comps[0], comps[1])
+                conf_list.append(conf_line)
+                conf_dict[dict_key] = conf_line
+                dict_key_last = dict_key
```
Comment (MKLeb) [inline salt/modules/postfix.py:87]:
```suggestion
        if re.search(SWWS, line):
            # This is the continuation of a logical line
            if not dict_key_last:
                raise Exception("Found continuation but no prior logical line")
            raw_line = line.strip()
            conf_dict[dict_key_last]["command"] = " ".join([conf_dict[dict_key_last]["command"], raw_line])
        elif re.search(r"^[^(#|$)]", line):
            # This is the start of a logical line
            comps = line.strip().split()
            conf_line = {
                "service": comps[0],
                "conn_type": comps[1],
                "private": comps[2],
                "unpriv": comps[3],
                "chroot": comps[4],
                "wakeup": comps[5],
                "maxproc": comps[6],
                "command": " ".join(comps[7:]),
            }
            dict_key = "{} {}".format(comps[0], comps[1])
            conf_list.append(conf_line)
            conf_dict[dict_key] = conf_line
            dict_key_last = dict_key
        else:
            # This is a comment, blank line, or other non-logical line
            conf_list.append(line)
```

File: salt/modules/postfix.py [hunk 2]
```diff
@@ -55,24 +55,36 @@ def _parse_master(path=MASTER_CF):
...
```
Comment (MKLeb) [inline salt/modules/postfix.py:88]:
```suggestion
        if not line.strip() or line.strip().startswith("#"):
            # Non-logical line
            conf_list.append(line)
        elif re.search(SWWS, line):
            # This is a continuation of a logical line, append to previous
            if not conf_list:
                conf_list.append(line)
            elif not dict_key_last:
                raise Exception("Found line continuation but no previous line")
            else:
                conf_dict[dict_key_last]["logical_line"] = " ".join(conf_dict[dict_key_last]["logical_line"], line)
        else:
            # This is the start of a logical line
            dict_key_last += 1
            tmp_dict = {"logical_line": line}
            conf_list.append(tmp_dict)
            conf_dict[dict_key_last] = tmp_dict
    
    # Now we can parse out the individual lines for the dict
    for tmp_key in conf_dict:
        logical_line_dict = conf_dict.pop(tmp_key)
        comps = logical_line_dict["logical_line"].strip().split()
        conf_line = {
            "service": comps[0],
            "conn_type": comps[1],
            "private": comps[2],
            "unpriv": comps[3],
            "chroot": comps[4],
            "wakeup": comps[5],
            "maxproc": comps[6],
            "command": " ".join(comps[7:]),
        }
        dict_key = "{} {}".format(comps[0], comps[1])
        # we want to use the same dict ref so it "alters" the list as well
        for key, value in conf_line.items():
            logical_line_dict[key] = value
        conf_dict[dict_key] = logical_line_dict
```

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**